R/R/response.R

```diff
@@ -222,59 +214,206 @@ robyn_response <- function(InputCollect = NULL,
+  } else {
+    if (grepl("last|all", metric_ds[1])) {
+      ## Using last_n as metric_ds range
+      if ("all" %in% metric_ds) metric_ds <- paste0("last_", length(all_dates))
+      last_n <- ifelse(grepl("_", metric_ds[1]), as.integer(gsub("last_", "", metric_ds)), 1)
+      metric_ds <- tail(all_dates, last_n)
+      metric_ds_loc <- which(all_dates %in% metric_ds)
+      metric_ds_val <- all_dates[metric_ds_loc]
+      rg <- v2t(range(metric_ds_val), sep = ":", quotes = FALSE)
+      if (length(metric_value_updated) == 1) metric_value_updated <- rep(metric_value_updated, last_n)
+      if (length(metric_value_updated) != last_n) {
+        stop("Input 'metric_value' must have length of 1 or same length as 'last_n'")
+      }
+    } else {
+      ## Using dates as metric_ds range
+      if (all(is.Date(as.Date(metric_ds, origin = "1970-01-01")))) {
+        metric_ds <- as.Date(metric_ds, origin = "1970-01-01")
+        if (length(metric_ds) == 1) {
+          ## Using only 1 date
+          if (all(metric_ds %in% all_dates)) {
+            metric_ds_val <- metric_ds
+            metric_ds_loc <- which(metric_ds %in% all_dates)
+            if (!quiet) message("Using ds '", metric_ds_val, "' as the response period")
+          } else {
+            metric_ds_loc <- which.min(abs(metric_ds - all_dates))
+            metric_ds_val <- all_dates[metric_ds_loc]
+            if (!quiet) warning("Input 'metric_ds' (", metric_ds, ") has no match. Picking closest date: ", metric_ds_val)
+          }
+        } else if (length(metric_ds) == 2) {
+          ## Using two dates as "from-to" date range
+          metric_ds_loc <- unlist(lapply(metric_ds, function(x) which.min(abs(x - all_dates))))
+          metric_ds_loc <- metric_ds_loc[1]:metric_ds_loc[2]
+          metric_ds_val <- all_dates[metric_ds_loc]
+          metric_ds_n <- length(metric_ds_val)
+          if (!quiet & !all(metric_ds %in% metric_ds_val)) {
+            warning(paste(
+              "At least one date in 'metric_ds' input do not match any date.",
+              "Picking closest dates for range:", paste(range(metric_ds_val), collapse = ":")
+            ))
+          }
+          rg <- v2t(range(metric_ds_val), sep = ":", quotes = FALSE)
+          if (!quiet) message("Using ds ", rg, " (", metric_ds_n, " period(s) included) as the response period")
+          if (length(metric_value_updated) == 1) {
+            metric_value_updated <- rep(metric_value_updated, metric_ds_n)
+          }
+        } else {
+          ## Manually inputting each date
+          if (all(metric_ds %in% all_dates)) {
+            metric_ds_val <- metric_ds
+            metric_ds_loc <- which(metric_ds %in% all_dates)
+            avlb_dates <- all_dates[(all_dates >= min(metric_ds_val) & all_dates <= max(metric_ds_val))]
+            if (length(metric_ds_loc) != length(avlb_dates) & !quiet)
+              warning(sprintf("There are %s skipping dates within your 'metric_ds' input",
```

Comment:
Should we allow skipping dates to be able to plot saturation curves for a channel in a date range (that contained zeroes - which can be taken out manually too)?

```
library(patchwork)
channel <- "facebook_S"
metric_values <- InputCollect$dt_input[[channel]]
non_adst <- robyn_response(
  InputCollect = InputCollect,
  OutputCollect = OutputCollect,
  select_model = select_model,
  media_metric = channel,
  metric_value = metric_values[metric_values > 0],
  # metric_ds = "all",
  quiet = FALSE
)
yes_adst <- robyn_response(
  InputCollect = InputCollect,
  OutputCollect = OutputCollect,
  select_model = select_model,
  media_metric = channel,
  metric_value = metric_values[metric_values > 0],
  metric_ds = InputCollect$dt_input$DATE[metric_values > 0],
  quiet = FALSE
)
non_adst$plot + yes_adst$plot
```